jax/_src/cudnn/fusion.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import jax
+from jax import core as jax_core
+from jax.interpreters import mlir
+from jax.interpreters.mlir import hlo
+from jax.interpreters.mlir import ir
+
+
+
+def _cudnn_fusion_impl(*args, jaxpr, **unused_kwargs):
+    del unused_kwargs
+    return jax_core.jaxpr_as_fun(jaxpr)(*args)
+
+
+def _custom_abstract_eval(*args, jaxpr, **unused_kwargs):
+    del unused_kwargs
+    del args
+    return jaxpr.out_avals
+
+
+cudnn_fusion_p = jax_core.Primitive("cudnn_fusion")
+cudnn_fusion_p.multiple_results = True
+cudnn_fusion_p.def_abstract_eval(_custom_abstract_eval)
+cudnn_fusion_p.def_impl(_cudnn_fusion_impl)
+
+
+def call_cudnn_fusion(f, *args, **kwargs):
+    """Creates a new cudnn_fusion corresponding to calling the given function f with args and kwargs."""
+    jaxpr, out_shapes = jax.make_jaxpr(
+        functools.partial(f, **kwargs), return_shape=True
+    )(*args)
+    flat_args = jax.tree.leaves(args)
+    out_tree = jax.tree.structure(out_shapes)
+    out_flat = cudnn_fusion_p.bind(*flat_args, name=f.__name__, jaxpr=jaxpr)
+    return jax.tree.unflatten(out_tree, out_flat)
+
+
+def _cudnn_fusion_stablehlo_lowering(
+    ctx,
+    *args,
+    name,
+    jaxpr,
+):
+    """Make cudnn_fusion which calls the implementation function.
+    Currently this leaks a CallOp since we're using the `core_call_lowering`
+    function, but this should get cleaned up by DCE easily.
+    """
+    impl = mlir.core_call_lowering(ctx, *args, name=name + ".impl", call_jaxpr=jaxpr)
+    call_op = impl[0].owner
+    called_fn = call_op.attributes["callee"]
+    cudnn_fusion = hlo.CustomCallOp(
+        [r.type for r in call_op.results],
+        call_op.operands,
+        call_target_name="__cudnn$fusion",
+        called_computations=ir.ArrayAttr.get([called_fn]),
+    )
+    return cudnn_fusion.results
+
+
+mlir.register_lowering(cudnn_fusion_p, _cudnn_fusion_stablehlo_lowering)
+
+
+def cudnn_fusion(f):
```

Comment:
Add a docstring with an example and make sure it end up in the JAX documentation?
Add the limitation in the docstring (don't support grad, vmap and for sharding, user must manually use shard_map)

Reply:
Added a docstring with an example. For user documentation, I prefer to make this available in JAX for evaluation first, then announce through documentation. For the limitations you named, 1) it seems they are common for all custom JAX primitives, 2) I'm not sure how fundamental are they here and whether we can't eventually remove them for this path.

Reply:
2) Yes, by default, all JAX new primitives doesn't support vmat grad and sharding.
But normal end user don't know that. So making the limitation clear help them. Otherwise, they will only find the limitation while trying it.

End user can by themself add a vmap/grad/shard_map if they want, even if the current PR doesn't make it automatic.
So we can add something like:

This implementation don't automatize vmap, grad and sharding. See custom_vmap, custom_vjp/jvp(https://jax.readthedocs.io/en/latest/jep/2026-custom-derivatives.html) and shard_map(https://jax.readthedocs.io/en/latest/notebooks/shard_map.html) to do it yourself.

This way, the user has all the info he needs without needing to deep-dive. (Except custom_vmap that isn't documented.)

Reply:
OK, then this can be put into documentation when we decide to start recommending this feature to end users.